File: python/open_guardrail/guards/reading_time.py
"""Estimate and limit reading time."""

import time
from typing import Optional

from open_guardrail.core import GuardResult


class _ReadingTime:
    def __init__(
        self,
        *,
        action: str = "block",
        max_minutes: float = 5.0,
        words_per_minute: int = 200,
    ) -> None:
        self.name = "reading-time"
        self.action = action
        self._max_minutes = max_minutes
        self._wpm = words_per_minute

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        words = [w for w in text.split() if w]
        word_count = len(words)
        minutes = round(word_count / self._wpm, 1)
        triggered = minutes > self._max_minutes
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="reading-time",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=(
                f"Reading time {minutes}min exceeds"
                f" max {self._max_minutes}min"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details={
                "estimated_minutes": minutes,
                "word_count": word_count,
                "max_minutes": self._max_minutes,
            },
        )


def reading_time(
    *,
    action: str = "block",
    max_minutes: float = 5.0,
    words_per_minute: int = 200,
) -> _ReadingTime:
    return _ReadingTime(
        action=action,
        max_minutes=max_minutes,
        words_per_minute=words_per_minute,
    )

File: packages/guards/src/reading-time.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

interface ReadingTimeOptions {
  action: 'block' | 'warn';
  maxMinutes?: number;
  wordsPerMinute?: number;
}

function estimateReadingTime(text: string, wpm: number): { minutes: number; words: number } {
  const words = text.split(/\s+/).filter((w) => w.length > 0).length;
  const minutes = Math.round((words / wpm) * 10) / 10;
  return { minutes, words };
}

export function readingTime(options: ReadingTimeOptions): Guard {
  const maxMinutes = options.maxMinutes ?? 5;
  const wpm = options.wordsPerMinute ?? 200;

  return {
    name: 'reading-time',
    version: '0.1.0',
    description: 'Estimate and limit reading time',
    category: 'content',
    supportedStages: ['output'],
    async check(
      text: string,
      _ctx: GuardContext,
    ): Promise<GuardResult> {
      const start = performance.now();
      const { minutes, words } = estimateReadingTime(text, wpm);
      const triggered = minutes > maxMinutes;

      return {
        guardName: 'reading-time',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: { estimatedMinutes: minutes, wordCount: words, maxMinutes },
      };
    },
  };
}
